lib/stellar/sep10.rb

```diff
@@ -0,0 +1,364 @@
+        )
+      end
+
+      if operation.body.value.data_value.unpack("m")[0].size !=  48
+        raise InvalidSep10ChallengeError.new(
+          "The transaction's operation value should be a 64 bytes base64 random string"
+        )
+      end
+
+      if !verify_tx_signed_by(tx_envelope: envelope, keypair: server)
+        raise InvalidSep10ChallengeError.new(
+          "The transaction is not signed by the server"
+        )
+      end
+
+      time_bounds = transaction.time_bounds
+      now = Time.now.to_i
+
+      if time_bounds.nil? || !now.between?(time_bounds.min_time, time_bounds.max_time)
+        raise InvalidSep10ChallengeError.new("The transaction has expired")        
+      end
+
+      # Mirror the return type of the other SDK's and return a string
+      client_kp = Stellar::KeyPair.from_public_key(client_account_id.ed25519!)
+
+      return envelope, client_kp.address
+    end
+
+
+    Contract(C::KeywordArgs[
+      challenge_xdr: String,
+      server: Stellar::KeyPair,
+      signers: SetOf[::Hash]
+    ] => C::SetOf[::Hash])
+    # Verifies that for a SEP 10 challenge transaction all signatures on the transaction are accounted for.
+    #
+    # A transaction is verified if it is signed by the server account, and all other signatures match a signer 
+    # that has been provided as an argument. Additional signers can be provided that do not have a signature, 
+    # but all signatures must be matched to a signer for verification to succeed. 
+    #
+    # If verification succeeds a list of signers that were found is returned, excluding the server account ID.
+    #
+    # @param challenge_xdr [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::Keypair] keypair for server's account.
+    # @param signers [SetOf[::Hash]] The signers of client account.
+    #
+    # @return [SetOf[::Hash]]
+    #
+    # Raises a InvalidSep10ChallengeError if:
+    #     - The transaction is invalid according to Stellar::SEP10.read_challenge_tx.
+    #     - One or more signatures in the transaction are not identifiable as the server account or one of the 
+    #       signers provided in the arguments.
+    def self.verify_challenge_tx_signers(
+      challenge_xdr:,
+      server:,
+      signers:
+    )
+      if signers.empty?
+        raise InvalidSep10ChallengeError.new("No signers provided.")
+      end
+
+      te, _ = read_challenge_tx(
+        challenge_xdr: challenge_xdr, server: server
+      )
+
+      # ignore non-G addresses
+      #
+      # We're also including the server in the list of signers past to verify_tx_signatures
+      # in order to verify that no signature is consumed more than once.
+      g_signers = Set.new
+      signers.each do |signer|
+        begin
+          Stellar::Util::StrKey.check_decode(:account_id, signer['key'])
+        rescue
+          next
+        else
+          g_signers.add(signer)
+        end
+      end
```

Comment:
I think this is a bit more intuitive for the Go SDK since `signerSummary` is a key -> weight map. When `VerifyChallengeTxSigners` returns a list of strings, `VerifyChallengeTxThreshold` can still easily get the weights using that map.

Since in this SDK we're using a hash containing both the key and the weight as separate pairs instead of mirroring `signerSummary`, I think it would be more intuitive for the user to be consistent with returning and passing the same structure for the signers in all functions.

Reply:
How does someone use the function to verify a challenge transaction is signed with a set of signers when weight doesn't matter? The default case with SEP-10 is that we're talking about weights and thresholds, but it's possible for a SEP-10 server to implement without a concern for thresholds and to ask for all signers to sign at transaction. I guess in that case they pass the hash without the weight and it works, is that right?

Assuming you keep the hash in this function, it still needs to deal with de-duplication of the signer.

Reply:
There are really two functions the user should use to verify sep 10 signers:

`verify_challenge_tx_signers`
This function is the one to use if you don't care about weights and thresholds. You could pass the set of hashes without the `'weight'` key and it would still work.

`verify_challenge_tx_threshold`
This function calls `verify_challenge_tx_signers` but also adds up the weights of the signers returned and checks the threshold.

Yes, I'll make sure the deduplication logic is still there.

Reply:
I went and looked at our other SDKs to get a refresh, and this is going to be different to the others. After thinking about it more, and writing out some different examples (see below) to see how the API fields, I don't see this being more convenient or intuitive for the user.

If a function asks for a list of signers and the user has only a list of signers (not weights), it won't be intuitive that the signers need to be wrapped in a hash that looks like this:
```ruby
signers = [
  { "key": "G..." },
  { "key": "G..." },
  { "key": "G..." }
]
_ = verify_challenge_tx_signers(challenge_xdr: tx, server: server_kp, signers: signers)
```
It will be easier to just pass in a list of signers:
```ruby
signers = [
  "G...",
  "G...",
  "G..."
]
_ = verify_challenge_tx_signers(challenge_xdr: tx, server: server_kp, signers: signers)
```
If a user does have the hashes from Horizon that contain both signer and weight and they need to pass in the list of signers, it's trivial at that point to do that using the `map` function, so it won't be inconvenient for the user:
```ruby
signer_details = [
  { "key": "G...", "weight": 2 },
  { "key": "G...", "weight": 2 },
  { "key": "G...", "weight": 3 }
]
signers = signer_details.map { |sd| sd["key"] }
_ = verify_challenge_tx_signers(challenge_xdr: tx, server: server_kp, signers: signers)
```

This will make this SDK deviate from our other SDKs (e.g. [JS SDK](https://stellar.github.io/js-stellar-sdk/Utils.html#.verifyChallengeTxSigners), [Go SDK](https://godoc.org/github.com/stellar/go/txnbuild#VerifyChallengeTxSigners), [Java SDK](https://stellar.github.io/java-stellar-sdk/org/stellar/sdk/Sep10Challenge.html#verifyChallengeTransactionSigners-java.lang.String-java.lang.String-org.stellar.sdk.Network-java.util.Set-), [Python SDK](https://stellar-sdk-zh-cn.readthedocs.io/en/latest/api.html#stellar_sdk.sep.stellar_web_authentication.verify_challenge_transaction_signers)), so in the absence of it being a significantly better API I think we should be consistent to make maintaining easier.

@stellar/horizon-committers (@ire-and-curses I saw you had already left some comments on this PR) Do you have an opinion on if this is worth deviating from other SDKs for?

Reply:
I don't feel too strongly either way, I see merits to both.

Quick question though: We're using `Set`s to parameters and returns values now. Should I revert back to using arrays?

Reply:
Good points on both sides I think. I see why it might be nice to use the same data structure without changes. Overall though IMO I lean to keeping this consistent with the other SDKs. I like the simpler function signature and it's not hard in Ruby to map from one to the other.

Reply:
>We're using `Set`s for parameters and return values now. Should I revert back to using arrays?

I'd keep using sets. We use them in the Java SDK. We probably would use them in the other languages if they existed built-in to their respective standard libraries. In languages they're available they simplify the deduplication and communicate intent better I think. In ruby it is very easy to convert from an array to a set and back again. A user can use the `.to_set` function on an array to get a set.